components/mission_key_management/client-reduced.c

```diff
@@ -0,0 +1,372 @@
+// Client state machine.  Each connection has a separate `struct client`
+// object, which tracks progress through the key request protocol for that
+// connection.
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "client.h"
+#include "mkm.h"
+
+#ifndef CN_ENV
+# include <sys/epoll.h>
+# include <unistd.h>
+# include <stdio.h>
+#else
+# include "cn_stubs.h"
+#endif
+
+
+uint32_t client_state_epoll_events(enum client_state state) {
+    switch (state) {
+        case CS_RECV_KEY_ID:
+            return EPOLLIN;
+        case CS_SEND_CHALLENGE:
+            return EPOLLOUT;
+        case CS_RECV_RESPONSE:
+            return EPOLLIN;
+        case CS_SEND_KEY:
+            return EPOLLOUT;
+        case CS_DONE:
+            return 0;
+
+        default:
+            // Unreachable
+            return 0;
+    }
+}
+
+struct client* client_new(int fd) 
+/*$
+// TODO: Doesn't handle the case where malloc fails 
+ensures 
+    take Client_out = Owned<struct client>(return);
+    Client_out.fd == fd; 
+    Client_out.pos == 0u8; 
+    ((i32) Client_out.state) == CS_RECV_KEY_ID;
+$*/
+{
+    // TODO: malloc spec assumes alloction always succeeds 
```

Comment:
The malloc spec can return a pointer that's valid ^ null using a predicate

Reply:
Yeah, it'd be good to have a standard way to do this. Do you have your experimental predicate hosted somewhere?

Reply:
See #146